### ISSUE TITLE
On the fly patched game does not run, pre-patched game does: order of mapper detection and patching is wrong

### ISSUE THREAD

**Issue by MBilderbeek** [openMSX/openMSX#1396]
I found this when playing with the IPS patch for Star Force as announced here: https://www.msx.org/news/en/star-force-smooth-scroll-enhanced

The original game is 32kB and the patch makes it something like 77kB (to be run with ASCII-8 mapper).

When you try to run it with openMSX using the on-the-fly patching you get this:
```
$ openmsx Star\ Force\ \(1985\)\ \(Hudson\)\ \(J\).rom -ips StarForce-ss-v0.0.ips
Fatal error: Error in "ROM" extension: Star Force: invalid rom size: must be smaller than or equal to 64kB and must be a multiple of 8kB.
```
However, when pre-patching the game (with a tool like uips), it runs fine (without specifying the mapper type). So for some reason, openMSX pads or doesn't complain about the size in this case, whilst in the on-the-fly patching case, it does complain or doesn't pad. 
Is the ROM type guessing done before the IPS patching perhaps?

This discrepancy is annoying and hard to explain to users. How can we fix this?

**Comment by m9710797** [comment]
IPS patching via openMSX or via an external patch-tool *should* behave the same, if not then there's a bug in one of the two tools (most likely in openMSX in this case).

I'll try to make time to investigate this later this week.

**Comment by mthuurne** [comment]
From the error message, it looks that the patching itself is not the problem, but the order in which mapper detection and patching are done, as Manuel suggested.

@MBilderbeek: If you force an ASCII-8 mapper on the command line, does the error go away?

**Comment by MBilderbeek** [comment]
@mthuurne Good point and indeed, it does work fine when I specify the ASCII8 mapper type. So, we can now confirm the mapper detection is at the wrong point in the pipeline.

**Comment by m9710797** [comment]
The code works as intended, though I agree that in this situation it's not ideal. Here's how it works:
See https://github.com/openMSX/openMSX/blob/master/src/memory/RomFactory.cc#L168

First we search the patched-sha1 in the rom database, in this case we don't find it. If not found then we search the pre-patched-sha1, we do find that one. Next we take the mapper-type that's specified in the database. But in this case that's the wrong type.
On the other hand if we pass an already patched rom to openMSX,  we (again) won't find this patched rom in the database. But now we cannot try the pre-patched rom. Thus we fall-back to "guessing" the mapper type. And in this case the guess is correct.

I think that in general trying the pre-patched-sha1 is a good idea: most IPS patches only patch small portions of the rom but keep the mapper-type the same. In this specific case the IPS patch does change the mapper-type (in fact the patch is more than twice the size of the original ROM). And then this heuristic goes wrong.

Any ideas for how to "solve" this? I'm inclined to just close this issue: it works as intended. And then yes: for IPS patches that change the rom-mapper-type you have to manually specify that type.

**Comment by mthuurne** [comment]
Maybe we could move the sanity checks out of the device constructor and call them as part of the auto-detection? Then if we'd auto-detect "plain ROM" based on the pre-patch SHA1, we'd verify that type and see its sanity checks fail and continue with auto-detection based on image heuristics. This would solve the case where a patch converts a small ROM into a MegaROM.

An alternative approach would be to run the image heuristics on the original and on the patched image and if the detected mapper type differs, don't fall back to the pre-patch SHA1, since it's likely that the mapper type has been patched.
